website/server/libs/groupTasks.js

```diff
@@ -1,24 +1,68 @@
 import * as Tasks from '../models/task';
+import {model as Groups} from '../models/group';
+import {model as Users} from '../models/user/index';
 
 const SHARED_COMPLETION = {
   default: 'recurringCompletion',
   single: 'singleCompletion',
   every: 'allAssignedCompletion',
 };
 
-async function _completeMasterTask (masterTask) {
-  masterTask.completed = true;
+// @TODO Group task scoring
+async function _completeOrUncompleteMasterTask (masterTask, completed) {
+  masterTask.completed = completed;
   await masterTask.save();
 }
 
-async function _deleteUnfinishedTasks (groupMemberTask) {
-  await Tasks.Task.deleteMany({
-    'group.taskId': groupMemberTask.group.taskId,
-    $and: [
-      {userId: {$exists: true}},
-      {userId: {$ne: groupMemberTask.userId}},
-    ],
-  }).exec();
+async function _updateAssignedUsersTasks (masterTask, groupMemberTask) {
+  if (groupMemberTask.type === 'todo') {
+    if (groupMemberTask.completed) {
+      // The task was done by one person and is removed from others' lists
+      await Tasks.Task.deleteMany({
+        'group.taskId': groupMemberTask.group.taskId,
+        $and: [
+          {userId: {$exists: true}},
+          {userId: {$ne: groupMemberTask.userId}},
+        ],
+      }).exec();
+    } else {
+      // The task was uncompleted by the group member and should be recreated for assignedUsers
+      let group = await Groups.findById(masterTask.group.id).exec();
+      let userList = [];
+      masterTask.group.assignedUsers.forEach(userId => {
+        let query = {_id: userId};
+        userList.push(query);
+      });
+      await Users.find({
+        $or: userList,
+      }).then(async assignedUsers => {
+        for (const assignedUser of assignedUsers) {
```

Comment:
the problem with this for is that it will run sequentially instead of in parallel.

My suggestion would be to only use await and not `then` to start with as it's very error prone

Reply:
Due to the `syncTask()` function returning a `Promise.all()` that includes a `masterTask.save()`, they must be done in sequence and waited on or mongoose will throw a `ParallelSaveError` because several asyncronous calls will be trying to update the same document. This sort of emulates the client making separate API calls to sync each assigned user. The alternative is to reimplement the `syncTask()` function.

I do not think the `group.save()` needs to happen after each user sync, though. That can likely come out of the loop and happen once.

Also, why not use `.then()`? In what ways is it error prone?

Reply:
Okay, I had an aha moment, but it is incomplete...
I do not totally understand the difference, but applied the pattern and it worked. I will take the question to Stackoveflow, since it is beyond the scope of this PR.

Thanks!

Reply:
> I do not think the group.save() needs to happen after each user sync, though. That can likely come out of the loop and happen once.

That would be great

> they must be done in sequence and waited on or mongoose will throw a ParallelSaveError because several asynchronous calls will be trying to update the same document

Database operations are usually quite slow, is there a way to only update each task once?

Reply:
The `group.taskSync()` function calls `.save()` on the `masterTask` passed to it. It does this because it is used (the only place, afaict) to assign a user to a task, and so will push the user into the `task.assignedUsers` array.

I see two option to optimize this for the purpose of being used both by the API and by this code (as well as hooking to saving the group task, which I'd like to do later):

1. Move `.push()`ing the `taskToSync` (aka `masterTask`, in this context) into the conditional that checks to see if the `user._id` is in the assigned users (line 1309). This strikes me as suboptimal, since it slightly complicates the logic of the function, and there could be some strange edge case if there is a race condition with un-assigning users and having assigned users un/completing tasks.
2. Move the assign/unassign into it's own function in `group.js` which saves the master task and then calls `syncTask()` and have the API handler call that. This seems far cleaner.

For optimization, I will take the second approach and push something tomorrow.

Reply:
I refactored the group model's `syncTask` to have a separate `linkTask` that will call it (and updated the API handler). This allowed me to effectively parallelize this section.

Thanks for the insights!!